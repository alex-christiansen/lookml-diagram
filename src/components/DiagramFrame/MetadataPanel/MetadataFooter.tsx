/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  Footer,
} from "@looker/components"
import styled from "styled-components"
import {METADATA_PANEL_PIXEL} from "../../../utils/constants"

export const MetadataFooter = styled(Footer as any)`
  width: ${METADATA_PANEL_PIXEL}px;
  right: -1px;
  position: absolute;
  bottom: 0px;
  border-top-color: transparent;
  box-shadow: 0px 1px 15px ${(props) => props.theme.colors.ui2};
  background-color: ${(props) => props.theme.colors.background};
`
