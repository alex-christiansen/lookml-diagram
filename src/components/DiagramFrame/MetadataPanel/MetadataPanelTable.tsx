
/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import React from 'react';
import {
  Flex,
  FlexItem,
  Text,
  theme
} from "@looker/components"
import styled from "styled-components";
import { LookmlObjectMetadata } from "../../types"
import { canGetDistribution, canGetTopValues } from "../../../utils/queries"
import { QueryChart } from "./QueryChart"

const MetadataRow = styled(Flex as any)`
  border-bottom: solid 1px ${(props) => props.theme.colors.ui2};
  width: 100%;
  padding-top: ${(props) => props.theme.sizes.xxsmall};
  padding-bottom: ${(props) => props.theme.sizes.xxsmall};
`
const KeyText = styled(Text as any)`
  font-size: ${(props) => props.theme.fontSizes.small};
  font-weight: 500;
`
const CodeText = styled(Text as any)`
  font-family: monospace;
  font-weight: 400;
`
const ValueText = styled(Text as any)`
  font-weight: 400;
`

const MetadataPanelTable: React.FC<{
  metadata: LookmlObjectMetadata,
  model?: any,
  explore?: any,
  field?: any
}> = ({
  metadata,
  model,
  explore,
  field
}) => {
	return (
  <Flex flexDirection="column">
    {metadata.fieldName && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>View Name</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <CodeText>{metadata.fieldName}</CodeText>
      </FlexItem>
    </MetadataRow>}
    {metadata.sqlTableName && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>SQL Table Name</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <CodeText>{metadata.sqlTableName}</CodeText>
      </FlexItem>
    </MetadataRow>}
    {metadata.label && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Label</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.label}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.viewLabel && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>View Label</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.viewLabel}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.groupLabel && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Group Label</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.groupLabel}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.fieldGroupLabel && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Field Group Label</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.fieldGroupLabel}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.labelShort && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Label Short</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.labelShort}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.accessFilters && metadata.accessFilters.length > 0 && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Access Filters</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
      {metadata.accessFilters.map((d,i) => {
        let spaceBelow = metadata.accessFilters.length > i+1
        return (
        <>
          <FlexItem>
            <KeyText variant="secondary">field: </KeyText>
            <ValueText>{d.field}</ValueText>
          </FlexItem>
          <FlexItem  pb={spaceBelow && "small"}>
            <KeyText variant="secondary">user_attribute: </KeyText>
            <ValueText>{d.user_attribute}</ValueText>
          </FlexItem>
        </>)
      })}
      </FlexItem>
    </MetadataRow>}
    {metadata.projectName && <MetadataRow>
      <FlexItem flexBasis="35%">
        <KeyText>Project Name</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.projectName}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {metadata.connectionName && <MetadataRow style={{borderBottom: `solid 1px ${theme.colors.ui2}`}}>
      <FlexItem flexBasis="35%">
        <KeyText>Connection Name</KeyText>
      </FlexItem>
      <FlexItem flexBasis="65%">
        <ValueText>{metadata.connectionName}</ValueText>
      </FlexItem>
    </MetadataRow>}
    {model && explore && field && <Flex pb="xxxlarge" mb="medium" flexDirection="column">
      <MetadataRow>
        <FlexItem flexBasis="35%">
          <KeyText>Distribution</KeyText>
        </FlexItem>
        <FlexItem flexBasis="65%">
          <QueryChart
          disabledText={
            "Distributions can only be shown for numeric dimensions on a view with a count measure."
          }
          enabled={canGetDistribution({ model, explore, field })}
          type={{
            type: "Distribution",
            model,
            explore,
            field
          }}
          />
        </FlexItem>
      </MetadataRow>
      <MetadataRow>
        <FlexItem flexBasis="35%">
          <KeyText>Values</KeyText>
        </FlexItem>
        <FlexItem flexBasis="65%">
          <QueryChart
          disabledText={
            "Values can only be shown for dimensions on a view with a count measure."
          }
          enabled={canGetTopValues({ model, explore, field })}
          type={{
            type: "Values",
            model,
            explore,
            field
          }}
          />
        </FlexItem>
      </MetadataRow>
    </Flex>}
  </Flex>
  )
}

export default MetadataPanelTable
