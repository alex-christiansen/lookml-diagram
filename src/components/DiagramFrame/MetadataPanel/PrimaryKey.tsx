import React from 'react'
import { getPkPath } from '../../../d3-utils/table-helpers'

export const PrimaryKey: React.FC<{
}> = ({
}) => {
  return (
    <svg className="lookml-diagram-primary-key" xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32" aria-labelledby="title">
 <title id="title">Lookml Diagram Primary Key Icon</title>
       <path d={getPkPath({primary_key: true})}/>
     </svg>
  )
}
