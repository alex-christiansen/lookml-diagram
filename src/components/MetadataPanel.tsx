import React, { useEffect } from 'react';
import {
  Badge,
  Box,
  ButtonTransparent,
  CodeBlock,
  Code,
  Chip,
  Flex,
  FlexItem,
  Text,
  SpaceVertical,
  Paragraph,
  Aside,
  Footer,
  Icon,
  Space,
  Tabs,
  Tooltip,
  Tab,
  TabList,
  TabPanel,
  TabPanels,
  theme,
  Heading
} from "@looker/components"
import styled from "styled-components";
import { ILookmlModel, ILookmlModelExplore } from "@looker/sdk"
import { ExternalLink } from "./ExternalLink"
import MetadataPanelTable from "./MetadataPanelTable"
import { METADATA_PANEL_PIXEL } from "../utils/constants"
import { LookmlObjectMetadata } from "./interfaces"
import { ILookmlModelExploreField, ILookmlModelExploreJoins } from '@looker/sdk/lib/sdk/3.1/models';
import { getFields } from '../utils/diagrammer'
import { exploreFieldURL } from '../utils/urls'
import { verticalAlign } from 'styled-system';

const MetadataInfoPanel = styled(Aside as any)`
  border-left: solid 1px ${(props) => props.theme.colors.ui2};
  overflow-y: auto;
  background-color: ${(props) => props.theme.colors.background};
  box-shadow: -10px 0px 20px 0px ${(props) => props.theme.colors.ui2};
  z-index: 0;
`
const MetadataRow = styled(Flex as any)`
  border-top: solid 1px ${(props) => props.theme.colors.ui2};
  width: 100%;
`
const KeyText = styled(Text as any)`
  font-size: ${(props) => props.theme.fontSizes.small};
  font-weight: 500;
`
const ValueText = styled(Text as any)`
  font-weight: 400;
`
const MetadataFooter = styled(Footer as any)`
  width: ${METADATA_PANEL_PIXEL}px;
  right: -1px;
  position: absolute;
  bottom: 0px;
  box-shadow: 0px 1px 15px ${(props) => props.theme.colors.ui2};
  background-color: ${(props) => props.theme.colors.background};
`

const LookmlCodeBlock = styled(CodeBlock as any)`
  background-color: ${(props) => props.theme.colors.neutralSubtle};
  color: ${(props) => props.theme.colors.keyInteractive};
`

const JoinPill = styled(Chip as any)`
  font-family: monospace; 
  border-radius: 20px;
`

const PillText:React.FC = ({children}) => {
  return <Code color="text3" fontSize="xsmall" lineHeight="medium">{children}</Code>
}

function getJoinCodeBlock(join: ILookmlModelExploreJoins) {
  let startLine = `join: ${join.name.toLowerCase()} {\n`
  let typeLine = join.type && `  type: ${join.type}\n`
  let relationLine = join.relationship && `  relationship: ${join.relationship}\n`
  let sqlLine = join.sql_on && `  sql_on: ${join.sql_on} ;;\n`
  let endLine = `}`
  return [startLine, typeLine, relationLine, sqlLine, endLine].filter(Boolean).join("")
}

function getFieldCodeBlock(field: ILookmlModelExploreField) {
  let startLine = `${field.category}: ${field.name.split(".")[1].toLowerCase()} {\n`
  let keyLine = field.primary_key && `  primary_key: yes\n`
  let typeLine = field.type && `  type: ${field.type}\n`
  let vfLine = field.value_format && `  value_format: ${field.value_format} ;;\n`
  let sqlLine = field.sql && `  sql: ${field.sql} ;;\n`
  let mapLayerLine = field.map_layer && field.map_layer.name && `  map_layer_name: ${field.map_layer.name}\n`
  let endLine = `}`
  return [startLine, keyLine, typeLine, vfLine, sqlLine, mapLayerLine, endLine].filter(Boolean).join("")
}

const MetadataPanel: React.FC<{
  explore: ILookmlModelExplore,
  selectionInfo: any,
  model: any,
}> = ({
  explore,
  selectionInfo,
  model
}) => {
  let metadata: LookmlObjectMetadata
  let field: ILookmlModelExploreField
  // console.log(explore, model)
  if (selectionInfo.lookmlElement === "explore") {
    // @ts-ignore
    let lookmlLink = explore.lookml_link
    metadata = {
      name: explore.name,
      description: explore.description,
      lookmlLink: lookmlLink,
      connectionName: explore.connection_name,
      label: explore.label,
      groupLabel: explore.group_label,
      projectName: explore.project_name,
      accessFilters: explore.access_filters,
    }
  } else if (selectionInfo.lookmlElement === "join") {
    let joinObj = explore.joins.filter((join: ILookmlModelExploreJoins) => {
      return join.name === selectionInfo.name
    })[0]
    // @ts-ignore
    let lookmlLink = explore.lookml_link
    let joinTypeRaw = joinObj.type ? joinObj.type : "left_outer"
    metadata = {
      name: joinObj.name,
      lookmlLink: lookmlLink,
      joinCode: getJoinCodeBlock(joinObj),
      joinIconType: joinTypeRaw.replace("_", "-"),
      joinType: joinTypeRaw.replace(/_/g, " ").toUpperCase(),
      joinRelationship: joinObj.relationship && joinObj.relationship.replace(/_/g, " ").toUpperCase(),
    }
  } else if (selectionInfo.lookmlElement === "dimension" || selectionInfo.lookmlElement === "measure") {
    let fields = getFields(explore.fields).filter((field: any) => {
      return field.name === selectionInfo.name || field.dimension_group === selectionInfo.name
    })
    let timeframes = fields.map((f: any) => {
      return f.type.includes("date_") && f.type.replace("date_", "")
    })
    field = fields[0]
    // @ts-ignore
    let lookmlLink = field.lookml_link
    metadata = {
      name: field.name.split(".")[1],
      fieldName: field.name.split(".")[0],
      lookmlLink: lookmlLink,
      fieldType: field.type.toUpperCase(),
      description: field.description,
      label: field.label,
      timeframes: !timeframes.includes(false) ? timeframes : [],
      // labelShort: field.label_short,
      fieldGroupLabel: field.field_group_label,
      valueFormat: field.value_format,
      userAttributeFilterTypes: field.user_attribute_filter_types,
      fieldSql: field.sql,
      primaryKey: field.primary_key,
      fieldCode: getFieldCodeBlock(field),
      fieldCategory: field.category.toUpperCase(),
    }
  } else if (selectionInfo.lookmlElement === "view") {
    // @ts-ignore
    let lookmlLink = explore.lookml_link
    metadata = {
      name: selectionInfo.name,
      lookmlLink: lookmlLink,
    }
  }
	return (
  <MetadataInfoPanel width={`${METADATA_PANEL_PIXEL}px`} px="medium" py="large">
    <SpaceVertical gap="xlarge">
      {/* NAME */}
      <Heading fontSize="xlarge" style={{fontWeight: 600}}>{metadata.name}</Heading>

      {/* PILLS */}
      <Space gap="xsmall">
        {metadata.joinType && 
        <Badge intent="neutral" size="medium">
          <Flex alignItems="center">
          <img
            src={
              `https://marketplace-api.looker.com/staging/app-assets/join-${metadata.joinIconType}.svg`
            }
            aria-hidden="true"
            style={{height: "24px"}}
          />
          <PillText>{metadata.joinType}</PillText>
          </Flex>
        </Badge>}
        {metadata.joinRelationship && <Badge intent="neutral" size="medium"><PillText>{metadata.joinRelationship}</PillText></Badge>}
        {metadata.fieldType && 
          <Badge intent="neutral" size="medium">
            <PillText>
              {metadata.fieldType}
            </PillText>
          </Badge>
        }
        {metadata.fieldCategory && 
          <Badge intent="neutral" size="medium">
            <PillText>
              {metadata.fieldCategory}
            </PillText>
          </Badge>
       }
        {metadata.primaryKey && 
          <Badge intent="neutral" size="medium">
            <Space gap="xsmall">
            <Icon name="Key" size="xsmall" />
            <PillText>Primary Key</PillText>
            </Space>
          </Badge>
        }
      </Space>

      {/* DESCRIPTION */}
      {metadata.description && (
        <Paragraph fontSize="small" color="text4">{metadata.description}</Paragraph>
      )}

      {/* JOIN CODE BLOCK */}
      {metadata.joinCode && (
        <LookmlCodeBlock>{metadata.joinCode}</LookmlCodeBlock>
      )}

      {metadata.fieldCode ?
      <>
      <Tabs>
        <TabList distribute>
          <Tab>Details</Tab>
          <Tab>Code</Tab>
        </TabList>
        <TabPanels pt={0}>
          <TabPanel>
            <MetadataPanelTable metadata={metadata} model={model} explore={explore} field={field} />
          </TabPanel>
          <TabPanel>
            <LookmlCodeBlock>{metadata.fieldCode}</LookmlCodeBlock>
          </TabPanel>
        </TabPanels>
      </Tabs>
      </> :
      <MetadataPanelTable metadata={metadata} /> }

    </SpaceVertical>
    <MetadataFooter
        borderTop={`1px solid ${theme.colors.ui2}`}
        pt="small"
        pb="small"
    >
      <Flex width="100%">
        <FlexItem flexBasis="60%"><ExternalLink target="_blank" href={metadata.lookmlLink}>
          <ButtonTransparent
            mr="xxxlarge"
            ml="small"
            iconBefore="LogoRings"
          >
            Go to LookML
          </ButtonTransparent>
        </ExternalLink></FlexItem>
        <FlexItem flexBasis="40%">{field && <ExternalLink
          target="_blank"
          href={exploreFieldURL(explore, field)}
        >
          <ButtonTransparent
            ml="xxxlarge"
            mr="xsmall"
            iconBefore="Explore"
          >
            Explore with Field
          </ButtonTransparent>
        </ExternalLink>}</FlexItem>
      </Flex>
    </MetadataFooter>
  </MetadataInfoPanel>
  )
}

export default MetadataPanel
