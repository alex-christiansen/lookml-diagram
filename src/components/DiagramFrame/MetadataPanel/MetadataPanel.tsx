/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react';
import {
  Badge,
  ButtonTransparent,
  Code,
  Flex,
  FlexItem,
  Heading,
  SpaceVertical,
  Paragraph,
  Icon,
  Space,
  Box,
  Tabs,
  Tab,
  TabList,
  TabPanel,
  TabPanels,
  theme
} from "@looker/components"
import { ILookmlModel, ILookmlModelExplore } from "@looker/sdk"
import { Explore, LogoRings } from "@looker/icons"
import { ILookmlModelExploreField, ILookmlModelExploreJoins } from '@looker/sdk/lib/sdk/3.1/models';

import { getFields } from '../../../utils/LookmlDiagrammer/'
import { exploreFieldURL } from '../../../utils/urls'
import { useExplore } from '../../../utils/fetchers'
import { METADATA_PANEL_PIXEL } from "../../../utils/constants"
import { LookmlObjectMetadata, SelectionInfoPacket } from "../../interfaces"
import { ExternalLink } from "../../ExternalLink"
import JoinIcon from "./JoinIcon"
import MetadataPanelTable from "./MetadataPanelTable"
import {MetadataInfoPanel} from "./MetadataInfoPanel"
import {MetadataFooter} from "./MetadataFooter"
import {LookmlCodeBlock} from "./LookmlCodeBlock"
import {getJoinMetadata, getFieldMetadata, getViewMetadata, getExploreMetadata} from "./utils"
import { PrimaryKey } from "./PrimaryKey"

export const PillText:React.FC = ({children}) => {
  return <Code color="text3" fontSize="xsmall" lineHeight="medium">{children}</Code>
}

export const MetadataPanel: React.FC<{
  explore: ILookmlModelExplore,
  selectionInfo: SelectionInfoPacket,
  model: ILookmlModel,
}> = ({
  explore,
  selectionInfo,
  model
}) => {
  let metadata: LookmlObjectMetadata
  let field: ILookmlModelExploreField
  // 'lookml_link' only exists on api response if user has "see_lookml"
  // permission. This is a requirement for using the extension. 
  // @ts-ignore
  let exploreLookmlLink = explore.lookml_link
  if (selectionInfo.lookmlElement === "explore") {
    metadata = getExploreMetadata(explore, exploreLookmlLink)
  } else if (selectionInfo.lookmlElement === "join") {
    let joinObj = explore.joins.filter((join: ILookmlModelExploreJoins) => {
      return join.name === selectionInfo.name
    })[0]
    metadata = getJoinMetadata(joinObj, exploreLookmlLink)
  } else if (selectionInfo.lookmlElement === "dimension" || selectionInfo.lookmlElement === "measure") {
    let fields = getFields(explore.fields).filter((field: any) => {
      return field.lookml_link === selectionInfo.link && (field.name === selectionInfo.name || (field.dimension_group === selectionInfo.grouped && field.name.includes(selectionInfo.name)))
    })
    field = fields[0]
    metadata = getFieldMetadata(fields, selectionInfo)
  } else if (selectionInfo.lookmlElement === "view") {
    let viewResponse = useExplore(explore.model_name, selectionInfo.name)
    metadata = getViewMetadata(viewResponse, exploreLookmlLink, selectionInfo)
  }
	return (
  <MetadataInfoPanel width={`${METADATA_PANEL_PIXEL}px`} px="medium" py="large">
    <SpaceVertical gap="large">
      {/* NAME */}
      <Heading fontSize="xlarge" style={{fontWeight: 500}}>{metadata.name}</Heading>

      {/* PILLS */}
      <Space gap="xsmall">
        {metadata.joinType && 
        <Badge intent="neutral" size="small">
          <Flex alignItems="center">
            <JoinIcon type={metadata.joinIconType}/>
            <PillText>{metadata.joinType}</PillText>
          </Flex>
        </Badge>}
        {metadata.joinRelationship && <Badge intent="neutral" size="small"><PillText>{metadata.joinRelationship}</PillText></Badge>}
        {metadata.fieldType && 
          <Badge intent="neutral" size="small">
            <PillText>
              {metadata.fieldType}
            </PillText>
          </Badge>
        }
        {metadata.fieldCategory && 
          <Badge intent="neutral" size="small">
            <PillText>
              {metadata.fieldCategory}
            </PillText>
          </Badge>
        }
        {metadata.primaryKey && 
          <Badge intent="neutral" size="small">
            <Space gap="xxsmall">
              <Icon icon={<PrimaryKey />} color={theme.colors.text3} pt="xxsmall"/>
              <PillText>Primary Key</PillText>
            </Space>
          </Badge>}
      </Space>

      {/* DESCRIPTION */}
      {metadata.description && (
        <Paragraph>{metadata.description}</Paragraph>
      )}

      {/* JOIN CODE BLOCK */}
      {metadata.joinCode && (
        <LookmlCodeBlock>{metadata.joinCode}</LookmlCodeBlock>
      )}

      <Box width="100%">
      {metadata.fieldCode ?
        <Tabs>
          <TabList distribute>
            <Tab>Details</Tab>
            <Tab>Code</Tab>
          </TabList>
          <TabPanels pt={0}>
            <TabPanel>
              <MetadataPanelTable metadata={metadata} model={model} explore={explore} field={field} />
            </TabPanel>
            <TabPanel>
              <LookmlCodeBlock>{metadata.fieldCode}</LookmlCodeBlock>
            </TabPanel>
          </TabPanels>
        </Tabs>
       :
      <MetadataPanelTable metadata={metadata} />
      }
      </Box>

    </SpaceVertical>
    <MetadataFooter
        borderTop={`1px solid ${theme.colors.ui2}`}
        pt="small"
        pb="small"
    >
      <Flex width="100%">
        <FlexItem flexBasis="60%"><ExternalLink target="_blank" href={metadata.lookmlLink}>
          <ButtonTransparent
            mr="xxxlarge"
            ml="small"
            iconBefore={<LogoRings />}
          >
            Go to LookML
          </ButtonTransparent>
        </ExternalLink></FlexItem>
        <FlexItem flexBasis="40%">{field && <ExternalLink
          target="_blank"
          href={exploreFieldURL(explore, field)}
        >
          <ButtonTransparent
            ml="xxxlarge"
            mr="xsmall"
            iconBefore={<Explore />}
          >
            Explore with Field
          </ButtonTransparent>
        </ExternalLink>}</FlexItem>
      </Flex>
    </MetadataFooter>
  </MetadataInfoPanel>
  )
}

