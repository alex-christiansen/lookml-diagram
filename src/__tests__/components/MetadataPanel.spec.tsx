/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react';
import {MetadataPanel} from '../../components/DiagramFrame/MetadataPanel/MetadataPanel'
import { shallow } from 'enzyme';
import {
  mockCurrentModel} from "../test_data";
// import 'jest-styled-components';

// jest.mock("@looker/components", () => ({
//   Divider: () => "Divider",
//   Flex: () => "Flex",
//   Aside: () => "Aside",
//   FlexItem: () => "FlexItem",
//   Card: () => "Card",
//   IconButton: () => "IconButton",
//   Badge: () => "Badge",
//   ButtonTransparent: () => "ButtonTransparent",
//   CodeBlock: () => "CodeBlock",
//   Code: () => "Code",
//   Heading: () => "Heading",
//   SpaceVertical: () => "SpaceVertical",
//   Paragraph: () => "Paragraph",
//   Footer: () => "Footer",
//   Icon: () => "Icon",
//   Space: () => "Space",
//   Tabs: () => "Tabs",
//   Tab: () => "Tab",
//   TabList: () => "TabList",
//   TabPanel: () => "TabPanel",
//   TabPanels: () => "TabPanels",
//   theme: {colors: {key:"rgb(45, 126, 234)"}, space: {large: "2em"}, fontSizes: {large: "2em"}, fontWeights: {normal: "1em"}}
// }))

// jest.mock('react', () => {
//   return {
//     useContext: jest.fn(() => {
//       return "useContext"
//     }),
//     createContext: jest.fn(() => {
//       return "createContext"
//     }),
//     createElement: jest.fn(() => {
//       return "createElement"
//     }),
//   }
// })

// jest.mock('@looker/extension-sdk-react', () => {
//   return {
//     ExtensionContext: jest.fn(() => {
//       return "ExtensionContext"
//     }),
//   }
// })

// jest.mock('styled-components', () => {
//   return {
//     SpinnerBlock: jest.fn(() => {
//       return "SpinnerBlock"
//     }),
//     Spinner: jest.fn(() => {
//       return "Spinner"
//     }),
//     DisabledText: jest.fn(() => {
//       return "DisabledText"
//     }),
//     _styledComponents: jest.fn(() => {
//       return "div"
//     }),
//     Aside: jest.fn(() => {
//       return "Aside"
//     }),
//   }
// })

// jest.mock("../../components/ExternalLink", () => ({
//   ExternalLink: () => "ExternalLink"
// }))

// jest.mock("../../components/DiagramFrame/MetadataPanel/JoinIcon", () => ({
//   JoinIcon: () => "JoinIcon"
// }))

// jest.mock("../../components/DiagramFrame/MetadataPanel/MetadataPanelTable", () => ({
//   MetadataPanelTable: () => "MetadataPanelTable"
// }))

// jest.mock("../../components/DiagramFrame/MetadataPanel/MetadataInfoPanel", () => ({
//   MetadataInfoPanel: () => "MetadataInfoPanel"
// }))

// jest.mock("../../components/DiagramFrame/MetadataPanel/MetadataFooter", () => ({
//   MetadataInfoPanel: () => "MetadataFooter"
// }))

// jest.mock("../../components/DiagramFrame/MetadataPanel/LookmlCodeBlock", () => ({
//   LookmlCodeBlock: () => "LookmlCodeBlock"
// }))

describe('<MetadataPanel />', () => {
  const basic = shallow(
  <MetadataPanel
    currentExplore={{joins:[{name:"join"}]}}
    selectionInfo={{lookmlElement: "join", name: "join"}}
    model={mockCurrentModel}
  />);
  it('should match the basic', () => {
    expect(basic).toMatchSnapshot();
  });
});
