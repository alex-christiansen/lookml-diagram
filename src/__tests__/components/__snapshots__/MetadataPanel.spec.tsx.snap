// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MetadataPanel /> should match the basic 1`] = `ShallowWrapper {}`;
